litespi/spi_nor_flash_module.py

```diff
@@ -180,113 +189,68 @@ class SpiNorFlashModule(metaclass=MetaSizes):
-                   (SpiNorFeatures.FEATURE_OCTAL not in cfg))
-        elif SpiNorFeatures.FEATURE_DUAL in cfg:
-            self.bus_width = 2
-            return ((SpiNorFeatures.FEATURE_SINGLE not in cfg) and
-                   (SpiNorFeatures.FEATURE_QPI not in cfg) and
-                   (SpiNorFeatures.FEATURE_OCTAL not in cfg))
-        elif SpiNorFeatures.FEATURE_QPI in cfg:
-            self.bus_width = 4
-            return ((SpiNorFeatures.FEATURE_SINGLE not in cfg) and
-                   (SpiNorFeatures.FEATURE_DUAL not in cfg) and
-                   (SpiNorFeatures.FEATURE_OCTAL not in cfg))
-        elif SpiNorFeatures.FEATURE_OCTAL in cfg:
-            self.bus_width = 8
-            return ((SpiNorFeatures.FEATURE_SINGLE not in cfg) and
-                   (SpiNorFeatures.FEATURE_DUAL not in cfg) and
-                   (SpiNorFeatures.FEATURE_QPI not in cfg))
-
-    def _configure_chip(self, cfg, erase_cmd, program_cmd, ddr_mode):
-        if cfg not in self.supported_modes:
-            raise ValueError("Configuration (%s) not supported in chip %s!" % (cfg, self.name))
-        assert self._configure_transfer_mode(cfg)
-
-        self.cmd_width = 1 # Currently we support only cmd_width=1
-        self.ddr = ddr_mode
-
-        # Find out what read command we can use
-        addr32 = ''
-        fast = ''
-        if SpiNorFeatures.FEATURE_4BA in cfg:
-            addr32 = '_4B'
-        if SpiNorFeatures.FEATURE_SINGLE in cfg:
-            if SpiNorFeatures.FEATURE_FAST_READ in cfg:
-                fast = '_FAST'
-            else:
-                self.dummy_bits = 0
-                self.read_opcode = getattr(SpiNorFlashOpCodes, "READ%s%s" %
-                          (fast, addr32))
-        else:
-            self.read_opcode = getattr(SpiNorFlashOpCodes, "READ_%d_%d_%d%s" %
-                          (self.cmd_width, self.addr_width, self.bus_width, addr32))
+        self.cmd_width = int(configs[0])
+        self.addr_width = int(configs[1])
+        self.bus_width = int(configs[2])
 
-        # Set rest of the commands
-        self.erase_opcode = erase_cmd
+    def _configure_chip(self, read_cmd, erase_cmd, program_cmd):
+        if read_cmd not in self.supported_opcodes:
+            raise ValueError("Read command (%s) not supported in chip %s!" % (str(read_cmd), self.name))
+        if program_cmd not in self.supported_opcodes:
+            raise ValueError("Read command (%s) not supported in chip %s!" % (str(read_cmd), self.name))
+
+        self._configure_transfer_mode(read_cmd)
+
+        # Set commands
+        self.read_opcode = read_cmd
         self.program_opcode = program_cmd
+        self.erase_opcode = erase_cmd
 
-    def __init__(self, spi_nor_cfg=SpiNorFeatures.FEATURE_SINGLE,
-                 erase_cmd=SpiNorFlashOpCodes.SE,
-                 program_cmd=SpiNorFlashOpCodes.PP,
-                 ddr_mode=False):
+    def __init__(self, read_cmd=SpiNorFlashOpCodes.READ_1_1_1,
```

Comment:
One complication is that Spi flash might need to initially start in the slow mode and then transition to the faster modes (mostly important for read).

Reply:
So we'll have `default_read_cmd` which configures chip into default mode and `read_cmds` which is a list of commands that can be used later on.